docs/propagation.md

```diff
@@ -0,0 +1,201 @@
+
+### Core Propagation Algorithm
+
+#### Propagate shardings along factors
+
+In Shardy, we have the hierarchy of tensors, dimensions, and factors. They represent data at different levels. A factor is a sub-dimension. It is an internal hierarchy used in sharding propagation. Each dimension may correspond to one or more factors. The mapping between dimension and factor is defined by OpShardingRule.
+
+**Shardy propagates sharding axes along factors instead of dimensions**. To do that, we have three steps as shown in the figure below
+
+1. Project DimSharding to FactorSharding
+2. Propagate sharding axes in the space of FactorSharding
+3. Projecte the updated FactorSharding to get the updated DimSharding
+
+#### Visualization of Sharding Propagation Along Factors
+
+We will use the following table to visualize the sharding propagation problem and algorithm.
+
+|  | F0 | F1 | F2 | Explicitly replicated axes |
+| :---- | :---- | :---- | :---- | :---- |
+| T0 |  |  |  |  |
+| T1 |  |  |  |  |
+| T2 |  |  |  |  |
+
+* Each column represents a factor. F0 means the factor with index 0. We propagate shardings along factors (columns).
+* Each row represents a tensor. T0 refers to the tensor with index 0. Tensors are all operands and results involved for a specific operation. The axes in a row cannot overlap. An axis (or sub-axis) cannot be used to partition one tensor many times. If an axis is explicitly replicated, we cannot use it to partition the tensor.
+
+Thus, each cell represents a factor sharding. A factor can be missing in partial tensors. The table for `C = dot(A, B)` is below. The highlighted cells imply that the factor is not in the tensor. For example, F2 is in T1 and T2, but not in T0.
+
+| `C = dot(A, B)` | F0 Batching dim | F1 Non-contracting dim | F2 Non-contracting dim | F3 Contracting dim | Explicitly replicated axes |
+| :---- | :---- | :---- | :---- | :---- | :---- |
+| T0 = A |  |  |  |  |  |
+| T1 = B |  |  |  |  |  |
+| T2 = C |  |  |  |  |  |
+
+#### Collect and propagate sharding axes
+
+We use a simple example shown below to visualize the propagation.
+
+|  | F0 | F1 | F2 | Explicitly replicated axes |
+| :---- | :---- | :---- | :---- | :---- |
+| T0 | “a” |  | “f” |  |
+| T1 | “a”, “b” | “c”, “d” | “g” |  |
+| T2 |  | “c”, “e” |  |  |
+
+**Step 1\.** Find axes to propagate along each factor (a.k.a. the (longest) compatible major sharding axes). For this example, we propagate \[“a”, “b”\] along F0, propagate \[“c”\] along F1, and propagate nothing along F2.
+
+**Step 2\.** Expand the factor shardings to obtain the following result.
+
+|  | F0 | F1 | F2 | Explicitly replicated axes |
+| :---- | :---- | :---- | :---- | :---- |
+| T0 | “a”, **“b”** | **“c”** | “f” |  |
+| T1 | “a”, “b” | “c”, “d” | “g” |  |
+| T2 | **“a”, “b”** | “c”, “e” |  |  |
+
+### Operations that are treated differently during propagation (**outdated**)
```

Comment:
This section is quite outdated, we handle those ops a bit differently in propagation.

The new logic is explained in these two places:
1. https://github.com/openxla/shardy/blob/main/docs/sdy_dialect.md#sdydata_flow_edge-sdydataflowedgeop
2. https://github.com/openxla/shardy/blob/main/docs/sdy_op_interfaces.md#shardabledataflowopinterface-shardabledataflowopinterface

We currently hard code the methods defined in the ShardableDataFlowOpInterface for StableHlo ops (while, case, optimization_barrier).

Let me know if you need more info on this.

Reply:
I am working on this and trying to fill in the blanks. I am probably wrong but best to ask now :) 

Originally, we said
```
When looking up the sharding of an operand of some operation op, the partitioner will “flow through” the optimization barrier. So below, when querying y_i, we will look at the sharding of x_i.
```

Looking at the data-flow edges documentation you've pointed out, my conclusion would be

````
An _owner_ is a user specified target of the data flow edge used by Shardy's propagation. The user can choose it arbitrarily but it needs to be static.

For example, given the `custom_op` defined below:

```c
  y_1, ..., y_n = custom_op (x_1, ..., x_n)
                  ((body_arg_1,..., body_arg_n) {
                    ...
                    return return_value_1, ..., return_value_n
                  })
```

This custom_op has two types for data flow edges: `n` edges each between `return_value_i` (sources) and `y_i` (targets) and `n` edges between `x_i` (sources) and `body_arg_i` (targets). **In this case, the edge owners are the same as the targets.**

When looking up the sharding of an operand of some operation `op`, the partitioner will "flow through" the optimization barrier. So below, when querying `return_value_i`, we will look at the sharding of the owner `y_i`.
````

Does this make sense?

Reply:
Ok I tried summarizing/updating the text, please let me know if you have other suggestions here.